text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,443 @@
+
+# Alternatives
+
+An alternative would be a reinterpretation of the user-provided parameters in the currently *infinite* cases. For example, the currently infinite `Range(0, 10, -1)` could be re-arranged to mean `Range(10, 0, -1)`. An argument against transforming Range(0, 10, -1) into Range(10, 0, -1) is that Range evaluates expressions using the provided parameters in their meaning of `min` and `max` or `start` and `end`. In conjunction with the step parameter there is a directionality (trajectory) which does not only control the numbers of the sequence but also the order within the sequence. Instead of `0, 1, .., 9` the Range would produce  `10, 9, .., 1` -- different numbers in a different order.
+Furthermore, having empty Ranges when using `Range` with calculated, non-literal bounds also has an *expressive* use (see CSV reader example in Motivation section) which the reinterpretation of the parameters to always return a non-empty Range would preclude.
+
+Another fair alternative is to leave everything as it is. Pony has been in fairly wide-spread use academically and to some degree even commercially, without anyone reporting the issues that for the author were the reason to start working on this RFC -- which was hanging (eventually running out of memory) Pony code. One could therefore speculate that the side effects and instances in which currently infinite Ranges are created are sufficiently documented and known by the majority of users of Pony.
+Nonetheless, as laid out in the summary and motivation sections, the use of the word Range suggests a behavior that is different from the current one. Therefore, if this RFC is not accepted, I'd suggest to better point out how Pony's use of `Range` is contrary to both the mathematical meaning and the use of iterator objects of that name in many other programming languages, and how this can lead to infinite loops over unintendedly infinite `Range` iterators. The current Range docstring does state this behavior and provides an example (infinite_range2), but this is possibly not prominent enough given the gravity of the discrepancy between reasonable expectation and behavior.
+
+>  If the `step` is not moving `min` towards `max` or if it is `0`,
+>  the Range is considered infinite and iterating over it
+>  will never terminate:
+> 
+> ```pony
+> let infinite_range1 = Range(0, 1, 0)
+> infinite_range1.is_infinite() == true
+> 
+> let infinite_range2 = Range[I8](0, 10, -1)
+> for _ in infinite_range2 do
+>   env.out.print("will this ever end?")
+>   env.err.print("no, never!")
+> end 
+> ```
```

Comment:
I do not think any of this is necessary for this RFC. None of it is relevant to Alternatives and by its own admission discussed in the Motivation already.

Reply:
>Your belief that there will be user code bugs is editorialzing at this point. Alternatives should be presented with the best >possible case for those alternatives such as "Leaving it as is means that code that might exist that I am unaware from >my search on GitHub would continue to work".

Sean reviewed an earlier draft and sort-of insisted on discussing the status quo.

Which parts would you remove?

Reply:
I highlighted the portion that I thought was not needed, but it does not look like that is visualized in the review properly. I was referring to this:

> Nonetheless, as laid out in the summary and motivation sections, the use of the word Range suggests a behavior that is different from the current one. Therefore, if this RFC is not accepted, I'd suggest to better point out how Pony's use of `Range` is contrary to both the mathematical meaning and the use of iterator objects of that name in many other programming languages, and how this can lead to infinite loops over unintendedly infinite `Range` iterators. The current Range docstring does state this behavior and provides an example (infinite_range2), but this is possibly not prominent enough given the gravity of the discrepancy between reasonable expectation and behavior.
> 
> >  If the `step` is not moving `min` towards `max` or if it is `0`,
> >  the Range is considered infinite and iterating over it
> >  will never terminate:
> > 
> > ```pony
> > let infinite_range1 = Range(0, 1, 0)
> > infinite_range1.is_infinite() == true
> > 
> > let infinite_range2 = Range[I8](0, 10, -1)
> > for _ in infinite_range2 do
> >   env.out.print("will this ever end?")
> >   env.err.print("no, never!")
> > end 
> > ```

I understand addressing the status quo from the purpose of explicitly stating what we lose by making these changes, but doubling back to the summary and motivation as you do here is not contributing towards addressing the status quo.

The status quo does more than merely not break existing code. You could also discuss how, after these changes, would have a new confusion around the reasonable assumption that a Range will produce values, but by adding empty Ranges this will no longer be true. The suggestion here is to move from having two categories (finite and infinite) of Ranges to having three categories (empty, finite, infinite). This can do more than break existing code, it could cause future confusion similar to what you had for expectations of Range to create this RFC. Your proposed solution to teaching this change is to modify the docstring...but you argue in this Alternatives section that the docstring was "possibly not prominent enough given the gravity of the discrepancy between reasonable expectation and behavior" -- something which is also possible for someone with an assumption that Ranges produce values.

Reply:
You raise good points. Should this potential confusion be added to Drawbacks then? Also, I will delete the passage you quoted..

Reply:
> You could also discuss how, after these changes, would have a new confusion around the reasonable assumption that a Range will produce values, but by adding empty Ranges this will no longer be true.

I only agree partially because Pony *had* empty Ranges but provided no means to test for that like .is_empty(). For example `Range(0, 0)` or `Range(10, 10, 10)` produce currently no iteration so that there should not be an user expectation for a Range to always produces values.

Reply:
I would add it to Drawbacks, yes.